cranelift/codegen/src/context.rs

```diff
@@ -159,18 +182,34 @@ impl Context {
 
         self.compute_domtree();
         self.eliminate_unreachable_code(isa)?;
-        if opt_level != OptLevel::None {
+
+        if !isa.flags().use_egraphs() && opt_level != OptLevel::None {
             self.dce(isa)?;
         }
 
         self.remove_constant_phis(isa)?;
 
-        if opt_level != OptLevel::None && isa.flags().enable_alias_analysis() {
+        if isa.flags().use_egraphs() {
+            log::debug!(
+                "About to optimize with egraph phase:\n{}",
+                self.func.display()
+            );
+            self.compute_domtree();
+            self.eliminate_unreachable_code(isa)?;
+            self.dce(isa)?;
+            self.remove_constant_phis(isa)?;
+            self.compute_loop_analysis();
+            let mut eg = FuncEGraph::new(&self.func, &self.domtree, &self.loop_analysis, &self.cfg);
+            eg.elaborate(&mut self.func);
+            log::debug!("After egraph optimization:\n{}", self.func.display());
+            log::info!("egraph stats: {:?}", eg.stats);
+        } else if opt_level != OptLevel::None && isa.flags().enable_alias_analysis() {
             self.replace_redundant_loads()?;
             self.simple_gvn(isa)?;
         }
 
-        isa.compile_function(&self.func, self.want_disasm)
+        self.optimized = true;
```

Comment:
I'm a little confused about this `optimized` flag.

I guess I'd like to at least see a comment about why `fn optimize` is `pub`, perhaps similar to the one on `compile_stencil`.

If `fn optimize` were not public, would it still be necessary to guard against it being called multiple times? Under what circumstances would it get called again?

Should the "already optimized" check be at the beginning of `optimize` instead of in `compile_stencil`?

I wonder, would it make sense to set `self.optimized` at the beginning of `fn optimize`? If an error occurs, I wouldn't expect the optimizer would work any better if you run it a second time. Maybe after failed optimization one might want to retry compilation without optimization?

Reply:
Yes, this was an attempt to make the API "smarter" in some sense, but I agree that the don't-re-optimize behavior is in the end just unnecessary complexity. I've marked `optimize()` as "only for testing" in the doc-comment and removed the `optimized` flag.